api/api/src/main/scala/uk/ac/wellcome/platform/api/rest/QueryParams.scala

```diff
@@ -48,23 +48,38 @@ trait QueryParamsUtils extends Directives {
   def decodeOneWithDefaultOf[T](default: T, values: (String, T)*): Decoder[T] =
     Decoder.decodeString.map { values.toMap.getOrElse(_, default) }
 
-  def decodeOneOfCommaSeparated[T](values: (String, T)*): Decoder[List[T]] =
+  def decodeOneOfCommaSeparated[T](values: (String, T)*): Decoder[List[T]] = {
+    val mapping = values.toMap
+    val validStrs = values.map(_._1).toList
     decodeCommaSeparated.emap { strs =>
-      val mapping = values.toMap
-      val results = strs.map { str =>
-        mapping
-          .get(str)
-          .map(Right(_))
-          .getOrElse(Left(str))
-      }
-      val invalid = results.collect { case Left(error) => error }
-      val valid = results.collect { case Right(value)  => value }
-      (invalid, valid) match {
-        case (Nil, results) => Right(results)
-        case (invalidValues, _) =>
-          Left(invalidValuesMsg(invalidValues, values.map(_._1).toList))
-      }
+      mapStringsToValues(strs, mapping).left
+        .map { invalidStrs =>
+          invalidValuesMsg(invalidStrs, validStrs)
+        }
     }
+  }
+
+  def decodeIncludesAndExcludes[T](
```

Comment:
Is it worth having a test for this as it's pretty self explanatory but has a few moving parts.

Reply:
Do you mean a test of just the decoder itself? We tend to just test the actual API functionality it provides (which I have done), but having specific unit tests for the decoders isn't a bad idea.

Reply:
Yea, for the decoder. 

If you give it rubbish, it gives you 🗑️ back etc.